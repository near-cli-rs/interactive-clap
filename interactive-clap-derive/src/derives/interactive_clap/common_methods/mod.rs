pub mod choose_variant;
pub mod fields_with_skip_default_input_arg;
pub mod from_cli_for_enum;
pub mod interactive_clap_attrs_context;
