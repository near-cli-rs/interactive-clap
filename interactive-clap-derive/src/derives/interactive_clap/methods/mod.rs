pub mod choose_variant;
pub mod cli_field_type;
pub mod fields_without_skip_default_input_arg;
pub mod fields_without_subcommand;
pub mod from_cli_for_enum;
pub mod from_cli_for_struct;
pub mod input_arg;
pub mod interactive_clap_attrs_context;
pub mod skip_interactive_input;
