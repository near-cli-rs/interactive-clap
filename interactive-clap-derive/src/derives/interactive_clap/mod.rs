extern crate proc_macro;

use proc_macro2::{Span, TokenStream};
use proc_macro_error::abort_call_site;
use quote::{quote, ToTokens};
use syn;

/// these are common methods, reused for both the [structs] and `enums` derives
pub(super) mod common_methods;

fn get_names(ast: &syn::DeriveInput) -> (&syn::Ident, syn::Ident) {
    let name = &ast.ident;
    let cli_name = {
        let cli_name_string = format!("Cli{}", name);
        syn::Ident::new(&cli_name_string, Span::call_site())
    };
    (name, cli_name)
}

pub fn impl_interactive_clap(ast: &syn::DeriveInput) -> TokenStream {
    let (name, cli_name) = get_names(ast);
    match &ast.data {
        syn::Data::Struct(data_struct) => {
            self::structs::token_stream(name, &cli_name, ast, &data_struct.fields)
        }
        syn::Data::Enum(syn::DataEnum { variants, .. }) => {
            let enum_variants = variants.iter().map(|variant| {
                let ident = &variant.ident;
                let mut attrs: Vec<proc_macro2::TokenStream> = Vec::new();
                if !&variant.attrs.is_empty() {
                    for attr in &variant.attrs {
                        if attr.path.is_ident("doc") {
                            attrs.push(attr.into_token_stream());
                        };
                        if attr.path.is_ident("cfg") {
                            for attr_token in attr.tokens.clone() {
                                match attr_token {
                                    proc_macro2::TokenTree::Group(group) => {
                                        if group.stream().to_string().contains("feature") {
                                            attrs.push(attr.into_token_stream());
                                        } else {
                                            continue;
                                        };
                                    }
                                    _ => {
                                        abort_call_site!("Only option `TokenTree::Group` is needed")
                                    }
                                }
                            }
                        };
                    }
                    match &variant.fields {
                        syn::Fields::Unnamed(fields) => {
                            let ty = &fields.unnamed[0].ty;
                            if attrs.is_empty() {
                                quote! {#ident(<#ty as interactive_clap::ToCli>::CliVariant)}
                            } else {
                                quote! {
                                    #(#attrs)*
                                    #ident(<#ty as interactive_clap::ToCli>::CliVariant)
                                }
                            }
                        }
                        syn::Fields::Unit => {
                            if attrs.is_empty() {
                                quote! {#ident}
                            } else {
                                quote! {
                                    #(#attrs)*
                                    #ident
                                }
                            }
                        }
                        _ => abort_call_site!(
                            "Only option `Fields::Unnamed` or `Fields::Unit` is needed"
                        ),
                    }
                } else {
                    match &variant.fields {
                        syn::Fields::Unnamed(fields) => {
                            let ty = &fields.unnamed[0].ty;
                            quote! { #ident(<#ty as interactive_clap::ToCli>::CliVariant) }
                        }
                        syn::Fields::Unit => {
                            quote! { #ident }
                        }
                        _ => abort_call_site!(
                            "Only option `Fields::Unnamed` or `Fields::Unit` is needed"
                        ),
                    }
                }
            });
            let for_cli_enum_variants = variants.iter().map(|variant| {
                let ident = &variant.ident;
                match &variant.fields {
                    syn::Fields::Unnamed(_) => {
                        quote! { #name::#ident(arg) => Self::#ident(arg.into()) }
                    }
                    syn::Fields::Unit => {
                        quote! { #name::#ident => Self::#ident }
                    }
                    _ => abort_call_site!(
                        "Only option `Fields::Unnamed` or `Fields::Unit` is needed"
                    ),
                }
            });

            let scope_for_enum = context_scope_for_enum(name);

            let fn_choose_variant =
                self::common_methods::choose_variant::fn_choose_variant(ast, variants);

            let fn_from_cli_for_enum =
                self::common_methods::from_cli_for_enum::from_cli_for_enum(ast, variants);

            quote! {
                #[derive(Debug, Clone, clap::Parser, interactive_clap::ToCliArgs)]
                pub enum #cli_name {
                    #( #enum_variants, )*
                }

                impl interactive_clap::ToCli for #name {
                    type CliVariant = #cli_name;
                }

                #scope_for_enum

                impl From<#name> for #cli_name {
                    fn from(command: #name) -> Self {
                        match command {
                            #( #for_cli_enum_variants, )*
                        }
                    }
                }

                #fn_from_cli_for_enum

                impl #name {
                    #fn_choose_variant

                    pub fn try_parse() -> Result<#cli_name, clap::Error> {
                        <#cli_name as clap::Parser>::try_parse()
                    }

                    pub fn parse() -> #cli_name {
                        <#cli_name as clap::Parser>::parse()
                    }

                    pub fn try_parse_from<I, T>(itr: I) -> Result<#cli_name, clap::Error>
                    where
                        I: ::std::iter::IntoIterator<Item = T>,
                        T: ::std::convert::Into<::std::ffi::OsString> + ::std::clone::Clone,
                    {
                        <#cli_name as clap::Parser>::try_parse_from(itr)
                    }
                }
            }
        }
        _ => abort_call_site!("`#[derive(InteractiveClap)]` only supports structs and enums"),
    }
}

/** This module describes [`crate::InteractiveClap`] derive logic in case when [`syn::DeriveInput`]
is a struct

The structure of produced derive output is as follows, where code blocks are generated by
submodules with corresponding names:

```rust,ignore
quote::quote! {
    #to_cli_trait_block
    #input_args_impl_block
    #to_interactive_clap_context_scope_trait_block
    #from_cli_trait_block
    #clap_for_named_arg_enum_block
}
```
*/
pub(crate) mod structs {
    pub(crate) mod to_cli_trait;

    mod input_args_impl;

    mod to_interactive_clap_context_scope_trait;

    mod from_cli_trait;

    mod clap_for_named_arg_enum;

    /// these are common field methods, reused by other [structs](super::structs) submodules
    pub(super) mod common_field_methods;

    /// returns the whole result `TokenStream` of derive logic of containing module
    pub fn token_stream(
        name: &syn::Ident,
        cli_name: &syn::Ident,
        ast: &syn::DeriveInput,
        fields: &syn::Fields,
    ) -> proc_macro2::TokenStream {
        let b1 = to_cli_trait::token_stream(name, cli_name, fields);
        let b2 = input_args_impl::token_stream(ast, fields);
        let b3 = to_interactive_clap_context_scope_trait::token_stream(ast, fields);
        let b4 = from_cli_trait::token_stream(ast, fields);
        let b5 = clap_for_named_arg_enum::token_stream(ast, fields);

        quote::quote! {
            #b1
            #b2
            #b3
            #b4
            #b5
        }
    }
}

fn context_scope_for_enum(name: &syn::Ident) -> proc_macro2::TokenStream {
    let interactive_clap_context_scope_for_enum = syn::Ident::new(
        &format!("InteractiveClapContextScopeFor{}", &name),
        Span::call_site(),
    );
    let enum_discriminants = syn::Ident::new(&format!("{}Discriminants", &name), Span::call_site());
    quote! {
        pub type #interactive_clap_context_scope_for_enum = #enum_discriminants;
        impl interactive_clap::ToInteractiveClapContextScope for #name {
                    type InteractiveClapContextScope = #interactive_clap_context_scope_for_enum;
                }
    }
}

#[cfg(test)]
pub(crate) mod to_cli_args_structs_test_bridge {
    struct Opts {
        name: syn::Ident,
        cli_name: syn::Ident,
        input_fields: syn::Fields,
    }
    fn prepare(ast: &syn::DeriveInput) -> Opts {
        let (name, cli_name) = super::get_names(ast);
        let input_fields = match &ast.data {
            syn::Data::Struct(data_struct) => data_struct.fields.clone(),
            syn::Data::Enum(..) | syn::Data::Union(..) => {
                unreachable!("stuct DeriveInput expected");
            }
        };
        Opts {
            name: name.clone(),
            cli_name,
            input_fields,
        }
    }

    pub fn partial_output(ast: &syn::DeriveInput) -> syn::Result<syn::DeriveInput> {
        let opts = prepare(ast);

        let (token_stream, _unused_byproduct) =
            super::structs::to_cli_trait::cli_variant_struct::token_stream(
                &opts.name,
                &opts.cli_name,
                &opts.input_fields,
            );
        syn::parse2(token_stream)
    }
}
