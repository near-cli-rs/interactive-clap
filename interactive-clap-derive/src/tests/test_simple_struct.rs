use super::pretty_codegen;
use crate::derives::interactive_clap::to_cli_args_structs_test_bridge;

#[test]
fn test_simple_struct() {
    let input = syn::parse_quote! {
        struct Args {
            age: u64,
            first_name: String,
            second_name: String,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_simple_struct_with_named_arg() {
    let input = syn::parse_quote! {
        struct Account {
            #[interactive_clap(named_arg)]
            field_name: Sender,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

/// this tested this problem https://github.com/near/near-cli-rs/pull/444#issuecomment-2631866217
#[test]
fn test_bug_fix_of_to_cli_args_derive() {
    let input = syn::parse_quote! {
        pub struct ViewAccountSummary {
            /// What Account ID do you need to view?
            account_id: crate::types::account_id::AccountId,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_flag() {
    let input = syn::parse_quote! {
        struct Args {
            /// Offline mode
            #[interactive_clap(long)]
            offline: bool
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
fn test_vec_multiple_opt() {
    let input = syn::parse_quote! {
        struct Args {
            #[interactive_clap(long_vec_multiple_opt)]
            pub env: Vec<String>,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}

#[test]
// testing correct panic msg isn't really very compatible with
// `proc-macro-error` crate
#[should_panic]
fn test_vec_multiple_opt_err() {
    let input = syn::parse_quote! {
        struct Args {
            #[interactive_clap(long_vec_multiple_opt)]
            pub env: String,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));
}

/// this test checks if doc comments are propagated up to `CliArgs` struct,
/// which has `clap::Parser` derive on it
///
/// also it checks that `#[interactive_clap(verbatim_doc_comment)]` attribute
/// gets transferred to `#[clap(verbatim_doc_comment)]` on `second_field` of
/// the same `CliArgs` struct
#[test]
fn test_doc_comments_propagate() {
    let input = syn::parse_quote! {
        struct Args {
            /// short first field description
            ///
            /// a longer paragraph, describing the usage and stuff with first field's
            /// awarenes of its possible applications
            #[interactive_clap(long)]
            #[interactive_clap(skip_interactive_input)]
            first_field: u64,
            /// short second field description
            ///
            /// a longer paragraph, describing the usage and stuff with second field's
            /// awareness of its possible applications
            #[interactive_clap(long)]
            #[interactive_clap(skip_interactive_input)]
            #[interactive_clap(verbatim_doc_comment)]
            second_field: String,
            /// short third field description
            ///
            /// a longer paragraph, describing the usage and stuff with third field's
            /// awareness of its possible applications
            #[interactive_clap(long)]
            #[interactive_clap(skip_interactive_input)]
            #[interactive_clap(verbatim_doc_comment)]
            third_field: bool,
        }
    };

    let interactive_clap_codegen = crate::derives::interactive_clap::impl_interactive_clap(&input);
    insta::assert_snapshot!(pretty_codegen(&interactive_clap_codegen));

    let step_two_input = to_cli_args_structs_test_bridge::partial_output(&input)
        .unwrap_or_else(|err| panic!("couldn't parse syn::DeriveInput: {:?}", err));

    let to_cli_args_codegen = crate::derives::to_cli_args::impl_to_cli_args(&step_two_input);
    insta::assert_snapshot!(pretty_codegen(&to_cli_args_codegen));
}
