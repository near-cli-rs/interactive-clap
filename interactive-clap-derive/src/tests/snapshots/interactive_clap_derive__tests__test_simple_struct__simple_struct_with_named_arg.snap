---
source: interactive-clap-derive/src/tests/test_simple_struct.rs
expression: pretty_codegen(&interactive_clap_codegen)
---
#[derive(Debug, Default, Clone, clap::Parser, interactive_clap::ToCliArgs)]
#[clap(author, version, about, long_about = None)]
pub struct CliAccount {
    #[clap(subcommand)]
    pub field_name: Option<ClapNamedArgSenderForAccount>,
}
impl interactive_clap::ToCli for Account {
    type CliVariant = CliAccount;
}
impl Account {
    pub fn try_parse() -> Result<CliAccount, clap::Error> {
        <CliAccount as clap::Parser>::try_parse()
    }
    pub fn parse() -> CliAccount {
        <CliAccount as clap::Parser>::parse()
    }
    pub fn try_parse_from<I, T>(itr: I) -> Result<CliAccount, clap::Error>
    where
        I: ::std::iter::IntoIterator<Item = T>,
        T: ::std::convert::Into<::std::ffi::OsString> + ::std::clone::Clone,
    {
        <CliAccount as clap::Parser>::try_parse_from(itr)
    }
}
impl From<Account> for CliAccount {
    fn from(args: Account) -> Self {
        Self {
            field_name: Some(args.field_name.into()),
        }
    }
}
impl Account {}
pub struct InteractiveClapContextScopeForAccount {}
impl interactive_clap::ToInteractiveClapContextScope for Account {
    type InteractiveClapContextScope = InteractiveClapContextScopeForAccount;
}
impl interactive_clap::FromCli for Account {
    type FromCliContext = ();
    type FromCliError = color_eyre::eyre::Error;
    fn from_cli(
        optional_clap_variant: Option<<Self as interactive_clap::ToCli>::CliVariant>,
        context: Self::FromCliContext,
    ) -> interactive_clap::ResultFromCli<
        <Self as interactive_clap::ToCli>::CliVariant,
        Self::FromCliError,
    >
    where
        Self: Sized + interactive_clap::ToCli,
    {
        let mut clap_variant = optional_clap_variant.clone().unwrap_or_default();
        let new_context_scope = InteractiveClapContextScopeForAccount {
        };
        let optional_field = match clap_variant.field_name.take() {
            Some(ClapNamedArgSenderForAccount::FieldName(cli_arg)) => Some(cli_arg),
            None => None,
        };
        match <Sender as interactive_clap::FromCli>::from_cli(
            optional_field,
            context.into(),
        ) {
            interactive_clap::ResultFromCli::Ok(cli_field) => {
                clap_variant
                    .field_name = Some(
                    ClapNamedArgSenderForAccount::FieldName(cli_field),
                );
            }
            interactive_clap::ResultFromCli::Cancel(optional_cli_field) => {
                clap_variant
                    .field_name = optional_cli_field
                    .map(ClapNamedArgSenderForAccount::FieldName);
                return interactive_clap::ResultFromCli::Cancel(Some(clap_variant));
            }
            interactive_clap::ResultFromCli::Back => {
                return interactive_clap::ResultFromCli::Back;
            }
            interactive_clap::ResultFromCli::Err(optional_cli_field, err) => {
                clap_variant
                    .field_name = optional_cli_field
                    .map(ClapNamedArgSenderForAccount::FieldName);
                return interactive_clap::ResultFromCli::Err(Some(clap_variant), err);
            }
        };
        interactive_clap::ResultFromCli::Ok(clap_variant)
    }
}
#[derive(Debug, Clone, clap::Parser, interactive_clap_derive::ToCliArgs)]
pub enum ClapNamedArgSenderForAccount {
    FieldName(<Sender as interactive_clap::ToCli>::CliVariant),
}
impl From<Sender> for ClapNamedArgSenderForAccount {
    fn from(item: Sender) -> Self {
        Self::FieldName(<Sender as interactive_clap::ToCli>::CliVariant::from(item))
    }
}
